pkg/network/server.go

```diff
@@ -130,6 +130,7 @@ type (
 		handshake  chan Peer
 		quit       chan struct{}
 		relayFin   chan struct{}
+		runFin     chan struct{}
```

Comment:
Much better. However, with this fix the following error occurs on Windows
```
2024-02-16T13:04:04.1864504Z panic: Log in goroutine after TestTryInitStateSync/module_already_initialized has completed: 2024-02-16T13:01:16.244Z	INFO	node started	{"blockHeight": 0, "headerHeight": 0}
2024-02-16T13:04:04.1864515Z 
2024-02-16T13:04:04.1864607Z goroutine 6763 [running]:
2024-02-16T13:04:04.1864830Z testing.(*common).logDepth(0xc007c211e0, {0xc000538000, 0x50}, 0x3)
2024-02-16T13:04:04.1865137Z 	C:/hostedtoolcache/windows/go/1.21.7/x64/src/testing/testing.go:1022 +0x6d4
2024-02-16T13:04:04.1865236Z testing.(*common).log(...)
2024-02-16T13:04:04.1865479Z 	C:/hostedtoolcache/windows/go/1.21.7/x64/src/testing/testing.go:1004
2024-02-16T13:04:04.1865746Z testing.(*common).Logf(0xc007c211e0, {0x1409d70b7, 0x2}, {0xc000406030, 0x1, 0x1})
2024-02-16T13:04:04.1866019Z 	C:/hostedtoolcache/windows/go/1.21.7/x64/src/testing/testing.go:1055 +0xa5
2024-02-16T13:04:04.1866445Z go.uber.org/zap/zaptest.testingWriter.Write({{0x140bc3c70?, 0xc007c211e0?}, 0x53?}, {0xc0002a7000, 0x51, 0x400})
2024-02-16T13:04:04.1866794Z 	C:/Users/runneradmin/go/pkg/mod/go.uber.org/zap@v1.24.0/zaptest/logger.go:130 +0x11e
2024-02-16T13:04:04.1867448Z go.uber.org/zap/zapcore.(*ioCore).Write(0xc0002b6240, {0x0, {0xc16bf5070e8f5b84, 0xc24fad41, 0x140fa1b40}, {0x0, 0x0}, {0x1409daaa0, 0xc}, {0x0, ...}, ...}, ...)
2024-02-16T13:04:04.1867806Z 	C:/Users/runneradmin/go/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/core.go:99 +0x193
2024-02-16T13:04:04.1868096Z go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc0002ce000, {0xc000394080, 0x2, 0x2})
2024-02-16T13:04:04.1868430Z 	C:/Users/runneradmin/go/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/entry.go:255 +0x2f0
2024-02-16T13:04:04.1868731Z go.uber.org/zap.(*Logger).Info(0xc000151600?, {0x1409daaa0, 0xc}, {0xc000394080, 0x2, 0x2})
2024-02-16T13:04:04.1869037Z 	C:/Users/runneradmin/go/pkg/mod/go.uber.org/zap@v1.24.0/logger.go:220 +0x65
2024-02-16T13:04:04.1869352Z github.com/nspcc-dev/neo-go/pkg/network.(*Server).Start(0xc000430c00)
2024-02-16T13:04:04.1869606Z 	D:/a/neo-go/neo-go/pkg/network/server.go:275 +0x29e
2024-02-16T13:04:04.1870020Z created by github.com/nspcc-dev/neo-go/pkg/network.startWithCleanup in goroutine 6762
2024-02-16T13:04:04.1870259Z 	D:/a/neo-go/neo-go/pkg/network/server_test.go:392 +0x91
2024-02-16T13:04:04.1870471Z FAIL	github.com/nspcc-dev/neo-go/pkg/network	3.648s
```

Reply:
As you see, the "faulty" line is `neo-go/neo-go/pkg/network/server.go:275 +0x29e`. This branch is already overwritten by the fresh changes, that's why it's important to investigate GH testing job failures after each push. However, this line is likely to be the log in the beginning of the server start function:
```
// Start will start the server and its underlying transport. Calling it twice
// is an error.
func (s *Server) Start() {
	if !s.started.CompareAndSwap(false, true) {
		s.log.Info("node server already started")
		return
	}
```
Or (which is more likely since there's no place in `TestTryInitStateSync` where server is double-started) the next log informing that server is started:
```
	s.log.Info("node started",
		zap.Uint32("blockHeight", s.chain.BlockHeight()),
		zap.Uint32("headerHeight", s.chain.HeaderHeight()))
```
I vote for second, so the reason of the test failure seems to be the following: there might be a situation when `s.tryInitStateSync()` is exiting earlier when this log is called, because `startWithCleanup` sends server run to a separate routine:
```
func startWithCleanup(t *testing.T, s *Server) {
	go s.Start()
```
So what we need here is to check that server is properly started, maybe with some `require.Eventually` line after `go s.Start()`.

Reply:
I'd suggest something like this, but we need @roman-khimov's approval for this since this code will be used by tests only:
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ git diff
diff --git a/pkg/network/server.go b/pkg/network/server.go
index fb72254a..e1bf2985 100644
--- a/pkg/network/server.go
+++ b/pkg/network/server.go
@@ -142,7 +142,8 @@ type (
 
                log *zap.Logger
 
-               started atomic.Bool
+               started        atomic.Bool
+               runLoopStarted atomic.Bool
        }
 
        peerDrop struct {
@@ -443,7 +444,11 @@ func (s *Server) run() {
                addrTimer        = time.NewTimer(peerCheckTime)
                peerTimer        = time.NewTimer(s.ProtoTickInterval)
        )
-       defer close(s.runFin)
+       s.runLoopStarted.Store(true)
+       defer func() {
+               s.runLoopStarted.Store(false)
+               close(s.runFin)
+       }()
        defer addrTimer.Stop()
        defer peerTimer.Stop()
        go s.runProto()
diff --git a/pkg/network/server_test.go b/pkg/network/server_test.go
index 2ca95885..d8623d45 100644
--- a/pkg/network/server_test.go
+++ b/pkg/network/server_test.go
@@ -390,6 +390,9 @@ func startTestServer(t *testing.T, protocolCfg ...func(*config.Blockchain)) *Ser
 
 func startWithCleanup(t *testing.T, s *Server) {
        go s.Start()
+       require.Eventually(t, func() bool {
+               return s.runLoopStarted.Load()
+       }, time.Second, 10*time.Millisecond)
        t.Cleanup(func() {
                s.Shutdown()
        })
```